lib/src/git.rs

```diff
@@ -987,73 +989,94 @@ fn update_git_head(
-        };
-        if !skip_reset {
-            git_repo.reset(new_git_commit.as_object(), git2::ResetType::Mixed, None)?;
-        }
+    // If the first parent of the working copy has changed, reset the Git HEAD.
+    if mut_repo.git_head() != first_parent {
+        update_git_head(
+            &git_repo,
+            gix::refs::transaction::PreviousValue::Any,
+            first_parent
+                .as_normal()
+                .map(|id| gix::ObjectId::from_bytes_or_panic(id.as_bytes())),
+        )?;
+        mut_repo.set_git_head_target(first_parent);
+    }
+
+    // If there is an ongoing operation (merge, rebase, etc.), we need to clean it
+    // up. This function isn't implemented in `gix`, so we need to use `git2`.
+    if git_repo.state().is_some() {
+        get_git_backend(mut_repo.store())
+            .ok_or(GitExportError::UnexpectedBackend)?
+            .open_git_repo()
+            .map_err(GitExportError::from_git)?
+            .cleanup_state()
+            .map_err(GitExportError::from_git)?;
+    }
+
+    // This is a way to find the tree ID associated with the raw Git commit, meaning
+    // it contains the ".jjconflict" trees as well. This is temporary; we just want
+    // to maintain the same behavior from git2.
+    let parent_tree_id = if first_parent_id == mut_repo.store().root_commit_id() {
+        mut_repo.store().empty_tree_id().clone()
     } else {
-        // Can't detach HEAD without a commit. Use placeholder ref to nullify the HEAD.
-        // We can't set_head() an arbitrary unborn ref, so use reference_symbolic()
-        // instead. Git CLI appears to deal with that. It would be nice if Git CLI
-        // couldn't create a commit without setting a valid branch name.
-        if mut_repo.git_head().is_present() {
-            match git_repo.find_reference(UNBORN_ROOT_REF_NAME) {
-                Ok(mut git_repo_ref) => git_repo_ref.delete()?,
-                Err(err) if err.code() == git2::ErrorCode::NotFound => {}
-                Err(err) => return Err(err),
-            }
-            git_repo.reference_symbolic("HEAD", UNBORN_ROOT_REF_NAME, true, "unset HEAD by jj")?;
-        }
-        // git_reset() of libgit2 requires a commit object. Do that manually.
-        let mut index = git_repo.index()?;
-        index.clear()?; // or read empty tree
-        index.write()?;
-        git_repo.cleanup_state()?;
+        TreeId::new(
+            git_repo
+                .find_commit(gix::ObjectId::from_bytes_or_panic(
+                    first_parent_id.as_bytes(),
+                ))
+                .map_err(GitExportError::from_git)?
+                .tree_id()
+                .map_err(GitExportError::from_git)?
+                .as_bytes()
+                .to_owned(),
+        )
+    };
+
+    let mut index = if &parent_tree_id == mut_repo.store().empty_tree_id() {
+        // If the tree is empty, gix fails to load the object, so we just use an empty
+        // index directly.
```

Comment:
nit: "gix fails" -> "gix can fail"?

It might be our fault that we don't write an empty tree object to the store even if we create a real Git commit with an empty tree.